tests/Microsoft.Identity.Test.Unit/ManagedIdentityTests/ImdsV2Tests.cs

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Net;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Client.AppConfig;
+using Microsoft.Identity.Client.ManagedIdentity;
+using Microsoft.Identity.Test.Common.Core.Mocks;
+using Microsoft.Identity.Test.Unit.Helpers;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Identity.Test.Unit.ManagedIdentityTests
+{
+    [TestClass]
+    public class ImdsV2Tests : TestBase
+    {
+        private readonly TestRetryPolicyFactory _testRetryPolicyFactory = new TestRetryPolicyFactory();
+
+        [TestMethod]
+        public async Task GetCsrMetadataAsyncSucceeds()
+        {
+            using (var httpManager = new MockHttpManager())
+            {
+                var handler = httpManager.AddMockHandler(MockHelpers.MockCsrResponse());
+
+                var managedIdentityApp = ManagedIdentityApplicationBuilder.Create(ManagedIdentityId.SystemAssigned)
+                    .WithHttpManager(httpManager)
+                    .WithRetryPolicyFactory(_testRetryPolicyFactory)
+                    .Build();
+
+                var miSource = await (managedIdentityApp as ManagedIdentityApplication).GetManagedIdentitySourceAsync().ConfigureAwait(false);
+                Assert.AreEqual(ManagedIdentitySource.ImdsV2, miSource);
+
+                Assert.IsTrue(handler.ActualRequestHeaders.Contains("Metadata"));
+                Assert.IsTrue(handler.ActualRequestHeaders.Contains("x-ms-client-request-id"));
+                Assert.IsTrue(handler.ActualRequestMessage.RequestUri.Query.Contains("api-version"));
+            }
+        }
+
+        [TestMethod]
+        public async Task GetCsrMetadataAsyncSucceedsAfterRetry()
+        {
+            using (var httpManager = new MockHttpManager())
+            {
+                // First attempt fails with INTERNAL_SERVER_ERROR (500)
+                httpManager.AddMockHandler(MockHelpers.MockCsrResponse(HttpStatusCode.InternalServerError));
+
+                // Second attempt succeeds
+                httpManager.AddMockHandler(MockHelpers.MockCsrResponse());
+
+                var managedIdentityApp = ManagedIdentityApplicationBuilder.Create(ManagedIdentityId.SystemAssigned)
+                    .WithHttpManager(httpManager)
+                    .WithRetryPolicyFactory(_testRetryPolicyFactory)
+                    .Build();
+
+                var miSource = await (managedIdentityApp as ManagedIdentityApplication).GetManagedIdentitySourceAsync().ConfigureAwait(false);
+                Assert.AreEqual(ManagedIdentitySource.ImdsV2, miSource);
+
+                const int NumRequests = 2; // initial probe request + 1 retry
+                int requestsMade = NumRequests - httpManager.QueueSize;
+                Assert.AreEqual(NumRequests, requestsMade);
```

Comment:
In MSAL.Net the tests fail if there are unused mocks left in the queue. This is not needed.

Reply:
Or you can simply validate that the QueueSize == 0

Reply:
Same comment for all the tests that follow this logic.